sparse_ho/criterion.py

```diff
@@ -20,178 +20,160 @@ class HeldOutMSE():
+        # TODO put the following in a callback function
+        self.mse_test = self.get_val_outer(
+            self.X_test, self.y_test, mask, dense)
         return val, grad
 
+    def proj_hyperparam(self, model, X, y, log_alpha):
+        return model.proj_hyperparam(
+            X[self.idx_train, :], y[self.idx_train], log_alpha)
+
 
 class HeldOutLogistic():
     """Logistic loss on held out data
     """
 
-    def __init__(self, X_val, y_val, model, X_test=None, y_test=None):
+    def __init__(self, idx_train, idx_val, X_test=None, y_test=None):
         """
         Parameters
         ----------
-        X_val : {ndarray, sparse matrix} of (n_samples, n_features)
-            Validation data
-        y_val : {ndarray, sparse matrix} of (n_samples)
-            Validation target
-        model: object of the class Model (e.g. Lasso or Sparse logistic regression)
+        idx_train: np.array
+            indices of the training set
+        idx_test: np.array
+            indices of the testing set
         X_test : {ndarray, sparse matrix} of (n_samples_test, n_features)
             Test data
         y_test : {ndarray, sparse matrix} of (n_samples_test)
             Test target
         """
-        self.X_val = X_val
-        self.y_val = y_val
+        self.idx_train = idx_train
+        self.idx_val = idx_val
         self.X_test = X_test
         self.y_test = y_test
-        self.model = model
 
         self.mask0 = None
         self.dense0 = None
         self.quantity_to_warm_start = None
         self.val_test = None
         self.rmse = None
 
-    def get_v(self, mask, dense):
-        temp = sigma(self.y_val * (self.X_val[:, mask] @ dense))
-        v = self.X_val[:, mask].T @ (self.y_val * (temp - 1))
-        v /= self.X_val.shape[0]
-        return v
-
-    def value(self, mask, dense):
-        val = np.sum(
-            np.log(1 + np.exp(-self.y_val * (self.X_val[:, mask] @ dense))))
-        val /= self.X_val.shape[0]
-        return val
-
-    def value_test(self, mask, dense):
-        if self.X_test is not None and self.y_test is not None:
-            self.val_test = np.sum(
-                np.log(1 + np.exp(-self.y_test * (self.X_test[:, mask] @ dense))))
-            self.val_test /= self.X_test.shape[0]
-        else:
-            self.val_test = None
-
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
+    @staticmethod
+    def get_val_outer(X, y, mask, dense):
+        if X is None or y is None:
+            return None
         else:
-            self.rmse = None
+            val = np.sum(
+                np.log(1 + np.exp(-y * (X[:, mask] @ dense))))
+            val /= X.shape[0]
+            return val
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
```

Comment:
I would rename this also. It's confusing (with get_val_outer), and it does something more than getting value.